Invite Brazilian Users to fill out a survey.

deBhal opened the pull request:
This PR adds a pop-up banner (currently an "Invitation", but needs a new name) to invite Brazilian users to fill out a survey about our recent focus on Brazil.  This invitation is intended to be run for a few weeks and then disabled.

There are a number of boxes to check:
- The invitation should only appear when the user locale is set to 'pt-BR'.
- Clicking on the Accept button should open the Brazilian survey in a new window.
- The invitation should only appear in the reader and my-sites sections (not in the editor or settings sections, for example)
- The invitation should not reappear after either the accept or dismiss buttons are pressed
- There should be tracks events for displaying, accepting and dismissing the invitation
- The invitation should look ok and work on mobile
- The invitation should not appear in production yet

To restore the invitation after clicking on a button, clear it out of the user preferences:
`require('lib/preferences/actions').set( 'dismissedBrazilianSurvey', null );`

It may help to turn debugging on for the invitation and for analytics:
`localStorage.debug = localStorage.debug + ', calypso:analytics, calypso:poll-invitation'`

deBhal commented:
I've added a globe icon into the invitation itself, but I need some help from a designer:
![account_settings_ _wordpress_com](https://cloud.githubusercontent.com/assets/5952255/13802269/d89327c8-eb85-11e5-8750-83322a254f42.jpg)
Firstly, colour and composition (12px margin on the right).  Are they ok?

I'd also like a sanity check on responsiveness.  `visibility: hidden; position: absolute;` works to hide the icon on smaller displays, but is that the right way to do it?
![configuracoes_da_conta_ _wordpress_com](https://cloud.githubusercontent.com/assets/5952255/13802383/dd4d1002-eb86-11e5-9d76-4dff2973ff50.jpg)


I've also got two other issues:
1) The icon scales at middling widths, how do I stop this?
![configuracoes_da_conta_ _wordpress_com](https://cloud.githubusercontent.com/assets/5952255/13802347/8925daae-eb86-11e5-811d-5bf0a8ee6cb0.jpg)
:

2) When it's book-ended by the icon and the buttons, the text looks oddly aligned vertically:
![configuracoes_da_conta_ _wordpress_com](https://cloud.githubusercontent.com/assets/5952255/13802360/a9f6ee76-eb86-11e5-8602-b7cc65db8735.jpg)

deBhal commented:
@aduth: Can I get your 2c on preferences/permanent storage here?  I've opted to use the preferencesData component, and punt on porting preferences to redux.  Does that seem appropriate?

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,145 @@
+import snakeCase from 'lodash/snakeCase';
+import includes from 'lodash/includes';
+import memoize from 'lodash/memoize';
+let debug = require( 'debug' )( 'calypso:poll-invitation' );
```
Two remarks:

- Generally we prefer to use `const` as much as possible over `let`, as the latter is a good indication that the value of the variable is reassigned at some point in time.
- The general approach we've taken with modules that export a function to be invoked is:

```js
/**
 * External dependencies
 */
import debugModule from 'debug';

/**
 * Internal dependencies
 */
import userModule from 'lib/user';

/**
 * Module variables
 */
const debug = debugModule( 'calypso:poll-invitation' );
const user = userModule();
```

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,145 @@
+	tracks.recordEvent( tracksEventName );
+}
+
+let recordEventOnce = memoize( recordEvent );
```
Consider instead: [`_.once`](https://lodash.com/docs#once)

deBhal replied:
Yeah, I had the same thought process.  `once()` is a better match with what this one function is doing, but `memoize()` is a better match for what I actually want, even though I only use the resulting function for the one thing.

I could easily go the other way if it reads badly.

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,145 @@
+
+		recordEventOnce( 'Displayed' );
+
+		let subComponents = {
```
`const` here instead of `let`. Do we really need to declare an object here, versus simply inlining the text within the rendered return value?

deBhal replied:
No, it's a separate object for historical reasons.

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,145 @@
+
+	dismiss: function() {
+		debug( 'dismiss' );
+		this.state.disabled = true;
```
Few remarks:

- Don't assign to `state` directly, use [`setState`](https://facebook.github.io/react/docs/component-api.html#setstate)
- Why is the debug and state assignment duplicated? Accidental copy-paste?

deBhal replied:
Rebase error :/

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,145 @@
+			return null;
+		}
+
+		if ( this.props.preferences[ _preferencesKey ] ) {
```
Small thing, but `_preferencesKey` is not a very clear variable name for what we're checking in preferences. Consider instead something like: `_dismissedPreferenceKey`

aduth commented:
>Can I get your 2c on preferences/permanent storage here?

I think it's fine, given that it's your only option for the moment. Good motivation to start migrating preferences to Redux though :smile:

deBhal commented:
Fixes are in, thanks for the review, @aduth!  I changed `memoize` to `once` as well.  If you had to stop and think about it, that's a bad sign.

folletto commented:
I checked it out: changed language to pt_br and the banner appeared in the Reader section as expected. I briefly tested functionality and seems ok.

Current screenshot:

![screen shot 2016-03-30 at 10 49 41](https://cloud.githubusercontent.com/assets/4389/14138480/5c977576-f665-11e5-9ce0-e54df7584c63.png)


I would suggest some small design tweaks:

![screen shot 2016-03-30 at 10 57 46](https://cloud.githubusercontent.com/assets/4389/14138638/440d1a82-f666-11e5-9b10-954e7fa8b4c2.png)

So:

1. Reverse button order (Calypso has the main action to the right)
2. Remove icon (we should use an illustration, but we don't have any good right now, and I think it's not adding much in this case
3. If possible align vertically the text so its two lines fit the height of the two buttons

deBhal commented:
@folletto Thanks for the review!  I've fixed up the button order and ditched the icon.

I wasn't entirely clear what you meant about aligning the buttons.  I've fixed up a margin from the column flow that was applying to the row flow, and that looks like this:

![aligned-no-extra-padding](https://cloud.githubusercontent.com/assets/5952255/14165708/2e763e08-f74f-11e5-8d4a-17d838eb8265.jpg)

Is that what you were referring to, or Did you mean to reduce the font size or pad out the buttons to make them the same height, like one of these? 
![aligned-smaller-font](https://cloud.githubusercontent.com/assets/5952255/14165756/b55292aa-f74f-11e5-8b3b-7ac454314d9e.jpg)

![aligned-extra-padding](https://cloud.githubusercontent.com/assets/5952255/14165761/bbf38326-f74f-11e5-8a77-2d12a49b8b44.jpg)

deBhal commented:
@aduth: are you happy for me to merge this with the issues you mentioned fixed?

folletto commented:
> I wasn't entirely clear what you meant about aligning the buttons. I've fixed up a margin from the column flow that was applying to the row flow, and that looks like this:

It's ok, I meant aligning the two lines to be vertically centered, but your first screenshot is good :+1: 

Tiny tiny detail: would be possible to reduce a bit the space between the two buttons (roughly half)? :)

aduth commented on config/development.json:
```diff
@@ -28,6 +28,7 @@
 	"features": {
 		"accept-invite": true,
 		"ad-tracking": false,
+		"brazil-survey-invitation":true,
```
When is this going live?

Also, minor, but space after colon.

deBhal replied:
Team global has a support rotation next week, which is a little awkward.

I'll discuss it with the team, but I'd love to have the option to launch today.

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,137 @@
+							Nós gostaríamos de lhe fazer 7 perguntas sobre o WordPress.com no Brasil.{ /* no translate(), pt-br only */ }
+						</p>
+						<div className="translator-invitation__actions">
+							<button
```
Using the [`<Button />` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/button) is recommended and tends to simplify the implementation.

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,137 @@
+	},
+
+	acceptButton: function() {
+		const url = 'https://href.li/?http://9372672.polldaddy.com/s/brazilian-portuguese-user-survey';
```
Can we assign this as `href` of the `<Button />`, making it a true link?

Also, do we want to be using the `href.li` service in redirecting the link?

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,137 @@
+	acceptButton: function() {
+		const url = 'https://href.li/?http://9372672.polldaddy.com/s/brazilian-portuguese-user-survey';
+		recordEvent( 'Clicked Accept Button' );
+		window.open( url );
```
e.g.

```jsx
<Button
	href="http://9372672.polldaddy.com/s/brazilian-portuguese-user-survey"
	target="_blank"
	onClick={ this.trackAndDismiss }
	primary />
```

aduth commented on client/layout/index.jsx:
```diff
@@ -131,6 +135,9 @@ Layout = React.createClass( {
 				</Welcome>
 				<TranslatorInvitation isVisible={ showInvitation } />
+				<PreferencesData>
+					<PollInvitation isVisible={ ! disablePollInvitation } section={ this.props.section } />
```
More we add here, more I wish these welcome prompts were split off into a separate component.

deBhal replied:
Yeah, I feel the same way.  We're only planning on running this one for a couple of weeks, but I think the next one needs to be a reusable component.

aduth commented on client/layout/poll-invitation/index.js:
```diff
@@ -0,0 +1,137 @@
+	displayName: 'PollInvitation',
+
+	propTypes: {
+		isVisible: React.PropTypes.bool
```
For consistency with DOM node attributes (e.g. `readOnly`, `disabled`), I personally discourage prefixing prop names with `is-` for indicating a boolean.

deBhal replied:
I like where you're coming from, but I'm going to punt this for a separate PR where I can change all three components in `client/layout/index.js` at once.

aduth replied:
>Out of curiosity - would you go with a boolean `visible`

Yeah, a boolean `visible` would be my preference. There's no helpers/wrappers for component visibility to my knowledge.

aduth commented:
Should the prompt be centered? At very large resolutions, the left-alignment looks a bit strange:

![image](https://cloud.githubusercontent.com/assets/1779930/14177998/c93ff952-f726-11e5-8191-964ec1d6e0ef.png)

folletto commented:
> Should the prompt be centered?

Well spotted, and yes. Should probably follow the reader in that sense.

deBhal commented:
Nice catch!  I've copied the styles from .main, and fixed up everything else ( except `isVisible` ).  It looks like this now:

![following_ _reader_ _wordpress_com](https://cloud.githubusercontent.com/assets/5952255/14197876/f27f47aa-f816-11e5-8d68-372a26b4e0f3.jpg)

I've also enabled it on horizon and staging.  I had a chat to Naoko, and we'd like to launch on Monday Morning Brazil time, so if we try this out on horizon and staging, I'll feel ok about enabling it on production during  our support week.

aduth commented:
This looks good to me :+1:

aduth commented:
Oh, please do squash plenty before merging.

deBhal commented:
> Oh, please do squash plenty before merging.
Haha, oh yes.  A *lot* of squashing.
